sql: add ability to not apply statement timeout to DECLARE CURSOR

otan opened the pull request:
Epic: none

Currently, `DECLARE CURSOR ... FETCH` cycles which take longer than the statement timeout would error. As a workaround, we introduce a `declare_cursor_statement_timeout_enabled` session variable which disables the timeout for `DECLARE CURSOR` only.

Works around #96322.

Release note (sql change): Introduce the
`declare_cursor_statement_timeout_enabled` session variable which disables statement timeouts during FETCH when using DECLARE CURSOR.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96607)
<!-- Reviewable:end -->

jordanlewis commented:
Rather than this session variable, can you:

1. Disable the statement timeout on the initial creation of the cursor
2. Pass in a context decorated with a cancellation with the session's actual statement timeout to each invocation of Fetch

?

otan commented:
> Pass in a context decorated with a cancellation with the session's actual statement timeout to each invocation of Fetch

how do we actually ensure the fetch cancels inside the IE and the like though?

we also want the cancel to apply during the DECLARE CURSOR planning stage. but i'm not sure how to get us to the behaviour described like https://github.com/cockroachdb/cockroach/issues/96322#issuecomment-1411952546.

(also i think this is safer to backport, we can work on doing this later)

jordanlewis commented:
Do we need a setting for this? It seems wrong to have a timeout on DECLARE ever. I think what we can do is close the internal executor if we see the context being cancelled when we check it in `ieResultChannel`.

otan commented:
I've done in a way that ticks the boxes over a riskier/new feature backport - it's new functionality controlled by a session variable. If we feel we can strongly about manipulating contexts to do this I can close this and do it that way.

jordanlewis commented:
I'm fine if you want to keep the session variable if you think it's really critical for safety. And maybe the context cancellation for FETCH isn't important since the IE is doing most of its work before the fetch comes in. I suppose the cursor blocks at the wrong spot. We could fix that but not in this PR.

I'll defer to @rafiss for an LGTM but I take back my objection.

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96607)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @otan)

___
*[`-- commits` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96607#-NNivX2M152qR5KdJCr-):*
hm i'd actually view this as a bug fix, since i can't see why anyone would want the previous behavior, and we've always tried to match postgres. but also i guess from a compatibility perspective, we are just introducing a new bug. wdyt?
___
*[`pkg/sql/sql_cursor.go` line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96607#-NNiwNRf2REfksriZc97:-NNiwNRf2REfksriZc98:b3hyrky) ([raw file](https://github.com/cockroachdb/cockroach/blob/31ec2bdcd06e47c3969a76dd94d7d27598c66832/pkg/sql/sql_cursor.go#L52)):*
> ```Go
> 
> 			sd := p.SessionData()
> 			if !p.SessionData().DeclareCursorStatementTimeoutEnabled {
> ```

nit: this new code could use a comment, and perhaps it should include the general idea of how want to fix this for real. or say if we never should bother to fix it
___
*[`pkg/sql/vars.go` line 2375 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96607#-NNivlgR-r3S2-JQ5NFV:-NNivlgR-r3S2-JQ5NFW:bfeydnd) ([raw file](https://github.com/cockroachdb/cockroach/blob/31ec2bdcd06e47c3969a76dd94d7d27598c66832/pkg/sql/vars.go#L2375)):*
> ```Go
> 			return formatBoolAsPostgresSetting(evalCtx.SessionData().DeclareCursorStatementTimeoutEnabled), nil
> 		},
> 		GlobalDefault: globalTrue,
> ```

i think we should set `Hidden: true` for this variable.

but also, is the backport going to default this to true? if so, i don't really see a reason for having the session variable.

(i guess the answer to bug fix vs sql change might affect what we do here)


<!-- Sent from Reviewable.io -->

otan committed:
sql: add ability to not apply statement timeout to DECLARE CURSOR

Epic: none

Currently, `DECLARE CURSOR ... FETCH` cycles which take longer than the
statement timeout would error. As a workaround, we introduce a
`declare_cursor_statement_timeout_enabled` session variable which
disables the timeout for `DECLARE CURSOR` only.

Release note (sql change): Introduce the
`declare_cursor_statement_timeout_enabled` session variable which
disables statement timeouts during FETCH when using DECLARE CURSOR.

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96607)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @rafiss)

___
*[`-- commits` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96607#-NNivX2M152qR5KdJCr-):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm i'd actually view this as a bug fix, since i can't see why anyone would want the previous behavior, and we've always tried to match postgres. but also i guess from a compatibility perspective, we are just introducing a new bug. wdyt?
</blockquote></details>

hmm, i see what you mean.
i'm of the opinion this is a workaround for an existing bug, which doesn't quite fix it.
i guess, from a users point of view, if they set a statement timeout they expect a timeout somewhere. silently ignoring it seems like a new bug? idk.

then again we're really only introducing this for AWS DMS to work. i don't know how often people use DECLARE CURSOR with a stmt_timeout.
___
*[`pkg/sql/vars.go` line 2375 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96607#-NNivlgR-r3S2-JQ5NFV:-NNixkPO2tHKhAOUXemi:b-k67fbz) ([raw file](https://github.com/cockroachdb/cockroach/blob/31ec2bdcd06e47c3969a76dd94d7d27598c66832/pkg/sql/vars.go#L2375)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think we should set `Hidden: true` for this variable.

but also, is the backport going to default this to true? if so, i don't really see a reason for having the session variable.

(i guess the answer to bug fix vs sql change might affect what we do here)
</blockquote></details>

hmm i still think it's a sql change. that said, i think it should be made available as we're likely going to be documenting this in any AWS DMS documentation.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96607)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @otan)

___
*[`-- commits` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96607#-NNivX2M152qR5KdJCr-):*
>  i don't know how often people use DECLARE CURSOR with a stmt_timeout.

to speculate: DMS was the first tool we wanted to support that requires DECLARE CURSOR, and it does use stmt_timeout with DECLARE cursor. so seems like a decent chance that using DECLARE CURSOR with stmt_timeout is common.

>  i guess, from a users point of view, if they set a statement timeout they expect a timeout somewhere. silently ignoring it seems like a new bug? idk.

yeah this is what i'm now wondering. how sure are we that completely ignoring the stmt_timeout in cursors won't just make the UX worse? presumably DMS added the timeout there for a reason. what if this results in behavior where things can hang forever?


<!-- Sent from Reviewable.io -->

otan commented:
*[`-- commits` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96607#-NNivX2M152qR5KdJCr-):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

>  i don't know how often people use DECLARE CURSOR with a stmt_timeout.

to speculate: DMS was the first tool we wanted to support that requires DECLARE CURSOR, and it does use stmt_timeout with DECLARE cursor. so seems like a decent chance that using DECLARE CURSOR with stmt_timeout is common.

>  i guess, from a users point of view, if they set a statement timeout they expect a timeout somewhere. silently ignoring it seems like a new bug? idk.

yeah this is what i'm now wondering. how sure are we that completely ignoring the stmt_timeout in cursors won't just make the UX worse? presumably DMS added the timeout there for a reason. what if this results in behavior where things can hang forever?
</blockquote></details>

wasn't the first (other GIS-y tools was what brought it to my knowledge), but yeah it's definitely one of them!

hmm, DMS is in reality doing a `SELECT * FROM db` and cursoring off that. it's "fine" if that takes longer. 

at the end of the day it's a somewhat crude "fix" and i guess the reason we're discussing whether it feels wrong is because it is wrong. i don't think fixing it proper is easy though, which is why i made this as a workaround. i'm ok to skip it but i think @awoods187 really wanted it.


<!-- Sent from Reviewable.io -->

awoods187 commented:
We need this functionality to enable AWS DMS CRDB to CRDB one-time replication--it would be very valuable to support this for 23.1. I think we should be generally okay taking on tech debt and revisiting that tech debt in the future. This fix could enable us to move faster now and we can fix it "properly" in 23.2. cc @ajstorm and @jordanlewis for opinions on the technical risk of taking on this tech debt

ajstorm commented:
Are we concerned that this won't work for DMS?  Or that the cluster setting will leak out and customers may use it outside of a DMS context? 

If it's the former, we may be able to run some testing to convince ourselves that in common cases, we're not exposing ourselves to infinite hangs.

rafiss commented:
@awoods187 My goal in pointing out my concerns is to ask how we communicate this change to our users. The specific questions are:
- Is it a change in functionality, or a bug fix?
- How do we tell users about the new bug that we are introducing intentionally?
- Should the new session variable be hidden?
- Should we even make this a setting?

I don't have a problem with merging this, but I am trying to make sure it is properly added to our release notes and that the knobs we introduce don't add more confusion. I am comfortable with tech debt, but with any tech debt there are ways to keep the debt in control and to cover our bases.

otan commented:
If it helps @awoods187, this is how I'm thinking about: I'm deliberately introducing a session variable that overrides the `statement_timeout` set for only `DECLARE CURSOR`  related statements.

As such:
* **Is it a change in functionality, or a bug fix?** new functionality gated by a session variable.
* **How do we tell users about the new bug that we are introducing intentionally?** we don't have to if it's a provided option that maintains the same behaviour by default.
* **Should the new session variable be hidden?** if we're documenting it in dms docs i don't think so
* **Should we even make this a setting?** yes because it does something "wrong".

There is a debate about whether we want to just not apply the `statement_timeout` by default for any `DECLARE CURSOR` statement. My feel is that that is a riskier change than this one as people may expect it to work how it works right now.  

For standard users, if they don't want `statement_timeout` on DECLARE CURSOR they should just set it to zero. Unfortunately, as we are discovering with DMS, we have no way to change the code in third party tools to do that.

The real fix here is to fix the `statement_timeout` for `DECLARE CURSOR` is much more involved (my guess is a week or two or more, but if you're nerd sniped into proving me wrong lmk).

otan commented:
bors r=rafiss
